common/authentication/aws/x509.go

```diff
@@ -0,0 +1,479 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/tls"
+	cryptoX509 "crypto/x509"
+	"errors"
+	"fmt"
+	"net/http"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	awssh "github.com/aws/rolesanywhere-credential-helper/aws_signing_helper"
+	"github.com/aws/rolesanywhere-credential-helper/rolesanywhere"
+	"github.com/aws/rolesanywhere-credential-helper/rolesanywhere/rolesanywhereiface"
+
+	cryptopem "github.com/dapr/kit/crypto/pem"
+	spiffecontext "github.com/dapr/kit/crypto/spiffe/context"
+	"github.com/dapr/kit/logger"
+	kitmd "github.com/dapr/kit/metadata"
+	"github.com/dapr/kit/ptr"
+)
+
+type x509 struct {
+	mu sync.RWMutex
+
+	wg      sync.WaitGroup
+	closeCh chan struct{}
+
+	logger              logger.Logger
+	Clients             *Clients
+	rolesAnywhereClient rolesanywhereiface.RolesAnywhereAPI // this is so we can mock it in tests
+	session             *session.Session
+	Cfg                 *aws.Config
+
+	chainPEM []byte
+	keyPEM   []byte
+
+	region          *string
+	TrustProfileArn *string        `json:"trustProfileArn" mapstructure:"trustProfileArn"`
+	TrustAnchorArn  *string        `json:"trustAnchorArn" mapstructure:"trustAnchorArn"`
+	AssumeRoleArn   *string        `json:"assumeRoleArn" mapstructure:"assumeRoleArn"`
+	SessionDuration *time.Duration `json:"sessionDuration" mapstructure:"sessionDuration"`
```

Comment:
I don't see any need to have a `SessionDuration` input.

Reply:
This is debatable. Community members get the most value add in having no session duration listed as then we handle the auto rotation of the session for them. However, I wanted to give the flexibility of if someone does want to have a one time session valid for up to the max 12 hours from AWS for a session then they could - maybe for a onetime component to do some checking on something? I am happy to remove, but again, kept to give flexibility here tbh. I suppose it would simplify things to remove too.

Reply:
Component lifecycle should be handled with the Component resource itself, rather than doing shenanigans with the underlying authentication. It is expected that any credential that the Component has be valid for the lifecycle of the Component. There isn't a world where a use would want to disable the auto-rotate, and we are creating foot guns by allowing it.